s2i/builder.go

```diff
@@ -0,0 +1,115 @@
+package s2i
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/url"
+	"os"
+
+	dockerClient "github.com/docker/docker/client"
+
+	"github.com/openshift/source-to-image/pkg/api"
+	"github.com/openshift/source-to-image/pkg/api/validation"
+	"github.com/openshift/source-to-image/pkg/build"
+	"github.com/openshift/source-to-image/pkg/build/strategies"
+	"github.com/openshift/source-to-image/pkg/scm/git"
+
+	fn "knative.dev/kn-plugin-func"
+	docker "knative.dev/kn-plugin-func/docker"
+)
+
+var (
+	// ErrRuntimeRequired indicates the required value of Function Runtime was not provided
+	ErrRuntimeRequired = errors.New("runtime is required to build")
+
+	// ErrRuntimeNotSupported indicates the given runtime is not (yet) supported
+	// by this builder.
+	ErrRuntimeNotSupported = errors.New("runtime not supported")
+)
+
+// DefaultBuilderImages for s2i builders indexed by Runtime Language
+var DefaultBuilderImages = map[string]string{
+	"node": "registry.access.redhat.com/ubi8/nodejs-16", // TODO: finalize choice and include version
```

Comment:
Right here is one of the downsides of using s2i instead of buildpacks. Buildpacks done properly allow the runtime version to be specified at build time. What happens if a user wants/needs to build a Node.js 14 application instead, because they have a dependency that depends on a deprecated Node.js API that was removed in 16. It's a hypothetical - something we should think about. But in all truth the problem lies with s2i architecture, and not with this line of code, IMO.

Reply:
I can't think of anything other than a brute-force approach where the latest node version is used as we do here, but there exists a mapping of major revisions to their builder image, and we accept an optional parameter where one can specify a Node version.

Note that the `--builder-image` flag is still available, so if the user is savvy about builder images, they can always provide that on the command line and it will take precedence over this default.

Reply:
> Note that the `--builder-image` flag is still available, so if the user is savvy about builder images, they can always provide that and it will take precedence over this default.

Yes and I think this also removes the need for an additional parameter. I was just pointing out the flaw in s2i.